### PR TITLE
keras 3.0.5

### PR DESCRIPTION
**Pull request by psteyer**
keras 3.0.5

**Destination channel:** Defaults

### Links

- [PKG-4035](https://anaconda.atlassian.net/browse/PKG-4035) 
- dev_url (pypi): https://github.com/keras-team/keras/tree/v3.0.5
- conda-forge: https://github.com/conda-forge/keras-feedstock/blob/main/recipe/meta.yaml
- pypi: https://pypi.org/project/keras/3.0.5

### Explanation of changes:

- updated recipe to version >3 using conda-forge
- removed all unnecessary files
- ~patch to set the default backend to `pytorch` (as per customer request)~ restore default behavior, set `KERAS_BACKEND` only for our tests
- ~script to remove leftover file `keras.json` before building (otherwise it will override the backend choice)~ removed this part as not needed anymore (we set the environment variable `KERAS_BACKEND` which overrides the setup file).

[PKG-4035]: https://anaconda.atlassian.net/browse/PKG-4035?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: recipe/meta.yaml
Comment (ViridianMelody) [inline recipe/meta.yaml]: This needs an update, as `Theano` is no longer supported.

File: recipe/meta.yaml
Comment (ViridianMelody) [inline recipe/meta.yaml:80]: Is `numpy` a valid backend? I was under the impression that only `jax`, `torch`, and `tensorflow` were supported.

File: recipe/patches/0001-backend_torch.patch
Comment (ViridianMelody) [inline recipe/patches/0001-backend_torch.patch]: Can I ask the reasoning behind using this patch instead of an activation script that sets `KERAS_BACKEND`?

**Review by ViridianMelody — APPROVED** [review]
Looks good to me. I had a few comments, but no significant blockers.

File: recipe/meta.yaml
Comment (psteyer) [reply recipe/meta.yaml:80]: I'm seeing it under the `backend` folder. So I believe it it might be.
https://github.com/keras-team/keras/tree/master/keras/src/backend/numpy.

But under the `README` I don't see it as an option.
https://github.com/keras-team/keras?tab=readme-ov-file#configuring-your-backend

File: recipe/meta.yaml
Comment (psteyer) [reply recipe/meta.yaml]: Updated the summary and description.

File: recipe/patches/0001-backend_torch.patch
Comment (psteyer) [reply recipe/patches/0001-backend_torch.patch]: I don't believe there was a specific reason. Just came to mind first. Is there a difference between using a patch or using an activation script in this case?

Comment (ViridianMelody) [reply recipe/patches/0001-backend_torch.patch]: Thinking through it a bit more, a patch might be preferable, since an activation script could overwrite the user's settings. I'm generally more inclined to leave it alone and let the user set it appropriately, but considering the request to change the default, this seems the least intrusive way to do so.

Comment (danpetry) [reply recipe/patches/0001-backend_torch.patch]: I think activation scripts are discouraged if there's another solution: https://docs.conda.io/projects/conda-build/en/latest/resources/activate-scripts.html#activate-scripts

File: recipe/meta.yaml
Comment (danpetry) [reply recipe/meta.yaml:80]: maybe there's a keras.backends property or similar you can read to make sure it's working right

**Review by danpetry — DISMISSED** [review]
I see why the default backend was changed (to update this without having to update tensorflow). However, this will change behaviour for people currently using the package. When they update, their backend will change from tensorflow to pytorch without them noticing. The backends are optional dependencies. Is it possible to keep the default backend at Tensorflow and just tell the customer to set KERAS_BACKEND themselves? That seems to be what was last communicated on the Jira ticket https://anaconda.atlassian.net/browse/PKG-4035?focusedCommentId=215906

Relatedly, the keras-base and keras-gpu packages have been removed, which are the ones that depend on tensorflow. I think they should be updated once we update tensorflow so the current users of those packages get the updates - these outputs could be left in, and commented out, with a comment explaining why.

**Comment by danpetry** [comment]
As discussed in slack, the other two packages were removed a long time ago, and tensorflow has a dependency on keras minor version which prevents auto-updates from happening with the backend changed. (the user has to manually install tensorflow too). However as @ViridianMelody pointed out, when tensorflow is updated to a supported version users will have to manually set the backend if they want to use tensorflow as before.

**Comment by danpetry** [comment]
IMHO the better way would be to remove the patch and tell snowflake to set the backend to pytorch themselves, but I'll leave that for discussion since the package needs to be delivered tomorrow

File: recipe/meta.yaml
Comment (lorepirri) [reply recipe/meta.yaml:80]: although not listed in the [README among the backends](https://github.com/keras-team/keras?tab=readme-ov-file#configuring-your-backend), it is checked as backend in `exports` and `contest` (`pytest` tests configuration):

- https://github.com/keras-team/keras/blob/v3.0.5/keras/backend/__init__.py#L40
- https://github.com/keras-team/keras/blob/v3.0.5/keras/backend/exports.py#L13
- https://github.com/keras-team/keras/blob/v3.0.5/conftest.py#L29

it is just a test btw, no harm is done.

File: recipe/patches/0001-backend_torch.patch
Comment (lorepirri) [reply recipe/patches/0001-backend_torch.patch]: I agree with @danpetry about the activation scripts. 

Previously, there were present but they were not utilized (not copied, not loaded), therefore we removed them. See https://github.com/AnacondaRecipes/keras-feedstock/pull/3/commits/d04fa2c920019ea0a06e62587f96aca41e6dab5e .

With this solution, we only set as default backend `torch`. The user can still override this choice using the environment variable, or the existing config.

Note: SF requested `keras` with `torch`.

**Comment by lorepirri** [comment]
> IMHO the better way would be to remove the patch and tell snowflake to set the backend to pytorch themselves, but I'll leave that for discussion since the package needs to be delivered tomorrow

@danpetry we can do that.

File: recipe/meta.yaml
Comment (skupr-anaconda) [inline recipe/meta.yaml:43]:
```suggestion
```
Seems `packaging` isn't required at runtime, see https://github.com/keras-team/keras/blob/v3.0.5/setup.py#L41-L48
It was added to the `master` branch and versions **>=3.3.0**

File: recipe/meta.yaml
Comment (skupr-anaconda) [inline recipe/meta.yaml]:
```suggestion
  license: Apache-2.0
  license_family: Apache
```
see https://github.com/keras-team/keras/blob/v3.0.5/LICENSE

Comment (lorepirri) [reply recipe/meta.yaml]: thank you @skupr-anaconda , will fix this

File: recipe/meta.yaml
Comment (lorepirri) [reply recipe/meta.yaml:43]: removed, thank you

Comment (skupr-anaconda) [reply recipe/meta.yaml:43]: Well, it's really needed:
```
  File "/croot/keras_1716277111998/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placeh/lib/python3.12/site-packages/keras/utils/torch_utils.py", line 3, in <module>

    from packaging.version import parse
ModuleNotFoundError: No module named 'packaging'
```

**Comment by lorepirri** [comment]
> When they update, their backend will change from tensorflow to pytorch without them noticing.

I mislead @psteyer with this suggestion:

1. The new users of Keras would see their default backend set to `torch` because `tensorflow` is not available. They can install this version, and they'll be able to use it with `torch`.
2. The previous users, would have their config in place with the correct backend, however, if they update, the correct version of `tensorflow` is not available. They should not update it.

However, your concern about the consistency of user experience made me think, thank you, and I'll restore the default behavior of `keras`.

File: recipe/meta.yaml
Comment (lorepirri) [reply recipe/meta.yaml:43]: I need to restore it: `packaging` is used in:
https://github.com/keras-team/keras/blob/v3.0.5/keras/utils/torch_utils.py#L3

Comment (lorepirri) [reply recipe/meta.yaml:43]: See log at https://cloud.prefect.io/anaconda/flow-run/c8a23c40-6ec4-4650-8952-ec8f5c4e9e36

Comment (lorepirri) [reply recipe/meta.yaml:43]: yup, saw it now ehehe

**Comment by lorepirri** [comment]
All concerns addressed. This is ready for another review. Thank you @danpetry and @skupr-anaconda !

File: recipe/meta.yaml
Comment (lorepirri) [reply recipe/meta.yaml:80]: added a comment for this in the recipe

File: recipe/meta.yaml
Comment (lorepirri) [reply recipe/meta.yaml:43]: restored, added a comment in the recipe

**Review by skupr-anaconda — APPROVED** [review]

File: recipe/meta.yaml
Comment (boldorider4) [inline recipe/meta.yaml]: I cannot find a specific requirement for this version of `jax` on upstream.
I think this line should be removed.

File: recipe/meta.yaml
Comment (boldorider4) [inline recipe/meta.yaml]: I am a little unsure whether for this specific version, since we do not provide `tensorflow` which was specified as run dependency, we should put `pytorch` in run.

Kind of a judgement call since in theory `keras` can still be installed and imported. Worst case scenario the presence of the backend is checked at runtime and an error (or warning) is thrown. but we did do this with `tensorflow` before.

Comment (lorepirri) [reply recipe/meta.yaml]: The three supported backends are in `run_constrained`, which is ok, they are optional (if none is installed, `keras` would be useless, but still a valid package).

You are right in mentioning doubts about the versions, which are:

```
jax =0.4.20
torch ~=2.1.0
tensorflow ~=2.16.1
```

which can be found here https://keras.io/getting_started/#tensorflow-compatibility

- we probably used `0.4.*` for `jax` to be less strict about it, and allow the `jax[cuda]` version to be installed (`0.4.24`) -> https://github.com/keras-team/keras/blob/v3.0.5/requirements-jax-cuda.txt .
- `pytorch >=2.1,<2.3` should be `pytorch >=2.1.0,<2.2.0` (expanding the tilde).
- `tensorflow >=2.15.0,<2.17.0a` should be `tensorflow >=2.16.1,<2.17.0` (expanding the tilde).

File: recipe/meta.yaml
Comment (lorepirri) [reply recipe/meta.yaml]: https://keras.io/getting_started/#tensorflow-compatibility

we probably used 0.4.* for `jax` to be less strict about it, and allow the `jax[cuda]` version to be installed (`0.4.24`) -> https://github.com/keras-team/keras/blob/v3.0.5/requirements-jax-cuda.txt .

File: recipe/meta.yaml
Comment (lorepirri) [inline recipe/meta.yaml]:
```suggestion
  run_constrained:
    # as suggested in https://keras.io/getting_started/#tensorflow-compatibility
    # tensorflow pinning expands ~=2.16.1
    - tensorflow >=2.16.1,<2.17.0a
    # should be 0.4.20, but this will allow the jax[cuda] version to be installed (0.4.24)
    - jax >=0.4.20, <0.5.0a
    # pytorch should have >=2.1.0,<2.2.0 (expanding ~=2.1.0) but in 
    # https://github.com/keras-team/keras/blob/v3.0.5/requirements.txt#L6
    # https://github.com/keras-team/keras/blob/v3.0.5/keras/utils/torch_utils.py#L156
    # a simple >=2.1.0 lower bound is used.
    - pytorch >=2.1.0
```

**Review by danpetry — APPROVED** [review]
Reviewed high-level/user facing aspects

File: recipe/meta.yaml
Comment (danpetry) [inline recipe/meta.yaml:17]: will this do anything? is `KERAS_BACKEND` a build time or runtime variable?

Comment (lorepirri) [reply recipe/meta.yaml:17]: It is a runtime variable as far as I can see. 

If it's not specified, the fallback is the default `tensorflow`, as expected.

I set it in the `build` section to `torch`, so that the first `import keras` won't fail while trying to import `tensorflow`.

File: recipe/meta.yaml
Comment (lorepirri) [reply recipe/meta.yaml]: @psteyer before merging, let's resolve this matter of which version pinnings we need.

Comment (lorepirri) [reply recipe/meta.yaml]: > since we do not provide tensorflow which was specified as run dependency, we should put pytorch in run.

Ah you meant in the previous version.

Good point.

we don't have the right version of tensorflow, and the default backend is still tensorflow (following the latest comments, I restored the default one to `tensorflow`).

Comment (lorepirri) [reply recipe/meta.yaml]: @skupr-anaconda what do you think?

Comment (lorepirri) [reply recipe/meta.yaml]: also on this point, @skupr-anaconda what is your opinion?

Comment (JeanChristopheMorinPerso) [inline recipe/meta.yaml]: This is should technically be `>=2.1.0,<2.2.0` if we follow what they say (`~=2.1`):
```python
>>> packaging.specifiers.Specifier('~=2.1.0').contains('2.0.1')
False
>>> packaging.specifiers.Specifier('~=2.1.0').contains('2.1.0')
True
>>> packaging.specifiers.Specifier('~=2.1.0').contains('2.1.10')
True
>>> packaging.specifiers.Specifier('~=2.1.0').contains('2.2.0')
False
```

But at the same time, their code says something else: https://github.com/keras-team/keras/blob/v3.0.5/keras/utils/torch_utils.py#L156 which says that it's compatible with `>=2.1` which is also true for the most recent keras version: https://github.com/keras-team/keras/blob/v3.3.3/keras/src/utils/torch_utils.py#L156.

File: recipe/meta.yaml
Comment (JeanChristopheMorinPerso) [inline recipe/meta.yaml]: This should be `>=2.16.1,<2.17.0a` because they say `tensorflow~=2.16.1` in https://keras.io/getting_started/#installing-keras-3.

```python
>>> packaging.specifiers.Specifier('~=2.16.1').contains('2.15.0')
False
>>> packaging.specifiers.Specifier('~=2.16.1').contains('2.16.0')
False
>>> packaging.specifiers.Specifier('~=2.16.1').contains('2.16.1')
True
>>> packaging.specifiers.Specifier('~=2.16.1').contains('2.16.6')
True
>>> packaging.specifiers.Specifier('~=2.16.1').contains('2.17.0')
False
```

File: recipe/meta.yaml
Comment (danpetry) [reply recipe/meta.yaml]: I think it should go in run_constrained. When we release tensorflow, this version of keras can be used with it without re-releasing keras

File: recipe/meta.yaml
Comment (danpetry) [reply recipe/meta.yaml:17]: does it import keras somewhere as part of the build process (i.e. when you run pip install?)

File: recipe/meta.yaml
Comment (lorepirri) [reply recipe/meta.yaml]: I've added `a` to my `>=2.16.1,<2.17.0` suggestion above, the rest seems ok.

File: recipe/meta.yaml
Comment (lorepirri) [reply recipe/meta.yaml]: also in the requirements, therefore the puzzle
https://github.com/keras-team/keras/blob/v3.0.5/requirements.txt#L6
following our past intention to be less restrictive with the pinnings, we should go for a simple `pytorch >=2.1.0`, without the upper bound.

Comment (lorepirri) [reply recipe/meta.yaml]: agree

Comment (lorepirri) [reply recipe/meta.yaml]: I've updated my proposal: https://github.com/AnacondaRecipes/keras-feedstock/pull/3#discussion_r1608214221

File: recipe/meta.yaml
Comment (lorepirri) [reply recipe/meta.yaml:17]: not in `setup.py` 

but in `pip_build.py` which is used to create a .whl on their ci:
```
        from keras.src.version import __version__  # noqa: E402
```
https://github.com/keras-team/keras/blob/b36f03731954037ea3b8a0f334368ceedad364b8/pip_build.py#L88

I don't think this will be a problem.

Comment (danpetry) [reply recipe/meta.yaml:17]: yeah, I don't think it's a problem to leave it in, but I don't think it does anything either. If that's right, could be misleading for future maintainers.

Comment (boldorider4) [reply recipe/meta.yaml:17]: Actually I would try building this without this variable.
Because if `keras` is not imported during the build step, setting this variable is pretty much useless.
If the build fails without this variable having been set, then yes, include it by all means.

I think where this variable should be set is an activation script.

Comment (danpetry) [reply recipe/meta.yaml:17]: > I think where this variable should be set is an activation script.

I guess you mean, if you want it to have effect at runtime, but we don't want to do that in this case? just to clarify

Comment (boldorider4) [reply recipe/meta.yaml:17]: yes, correct, but my understanding is that we want to have PyTorch as the only backend available for now, until `tensorflow` is updated.
In this case, `torch` should also be a `run` dependency of course.

Comment (JeanChristopheMorinPerso) [reply recipe/meta.yaml:17]: We support both pytorch and jax. Pytorch should not be a runtime dependency.

File: recipe/meta.yaml
Comment (boldorider4) [reply recipe/meta.yaml]: I see, great catch in this case!

File: recipe/meta.yaml
Comment (boldorider4) [reply recipe/meta.yaml]: I think in this case we normally would write
`- jax >=0.4.20, <0.5.0a`

File: recipe/meta.yaml
Comment (danpetry) [reply recipe/meta.yaml:17]: It was my understanding that we had [come to a decision](https://github.com/AnacondaRecipes/keras-feedstock/pull/3#issuecomment-2121885940) that we weren't swapping the default backend to pytorch, because it would change behaviour for users, rather that we were telling snowflake to configure the backend themselves.

I think changing the default backend is not the way to go, considering the non-snowflake users. I get that there's a hard pin in the tensorflow feedstock on keras, so users' environments won't get updated automatically. However, in the future when we release tensorflow, the default backend will still be pytorch. It means users who would want to update keras and continue using it with tensorflow will now have to configure it to use tensorflow => general confusion ensues. We should be really careful with these kinds of changes in user experience because it can break things for them.

If snowflake are really requiring the default to to become pytorch, and setting an environment variable is not acceptable for some reason, we could consider making them a special build for their channel, maybe?

Comment (danpetry) [reply recipe/meta.yaml:17]: and connectedly, all backends should be run_constrained because they're optional. unless it's a special build. (a package called `keras-torch` maybe)

Comment (danpetry) [reply recipe/meta.yaml:17]: I also wonder what would be the behaviour of the keras that tensorflow depends on if pytorch is in the same environment

Comment (danpetry) [reply recipe/meta.yaml:17]: For background, keras is a very popular AI package and this has the potential to have a significant impact

Comment (danpetry) [reply recipe/meta.yaml:17]: well, we can always re-release keras with the default backend changed and the build number bumped when we release tensorflow, I guess, to stop such breakage. Seems unnecessary though.

File: recipe/meta.yaml
Comment (lorepirri) [reply recipe/meta.yaml]: updated, thank you

File: recipe/meta.yaml
Comment (lorepirri) [reply recipe/meta.yaml:17]: thank you for helping here @danpetry , I just need `KERAS_BACKEND=torch` for the first `import keras` in the test section, and not a build time.

The default backend is still `tensorflow`, but I can't run tests with that, I need to run them on `torch`.

Do we have a way to set this variable just right before the tests start? Otherwise, I will remove the `imports` section and use a `KERAS_BACKEND=torch python -c "import keras;"` command under `commands`.

Comment (danpetry) [reply recipe/meta.yaml:17]: the test commands are just bash commands, so you can `export KERAS_BACKEND=torch` before any of the tests

Comment (lorepirri) [reply recipe/meta.yaml:17]: the `imports` section

**Review by lorepirri — CHANGES_REQUESTED** [review]
pinnings and tests need to be updated

File: recipe/meta.yaml
Comment (lorepirri) [reply recipe/meta.yaml]: committed

File: recipe/meta.yaml
Comment (lorepirri) [reply recipe/meta.yaml:17]: moved `tests/imports` section to the `test/commands` section

**Comment by lorepirri** [comment]
pinning and tests updated.

I think I've resolved all the comments.

Waiting for other reviews.

I will merge before the end of the week.

File: recipe/meta.yaml
Comment (danpetry) [reply recipe/meta.yaml:17]: ah I see. right yeah

File: recipe/patches/0001-backend_torch.patch
Comment (danpetry) [reply recipe/patches/0001-backend_torch.patch]: resolved, default backend kept as upstream (tensorflow).

**Review by lorepirri — APPROVED** [review]
No other comments received. Ready to merge.

**Pull request merged**
